File: go/spore-core/verifier/verifier_fixture_replay_test.go
package verifier

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// Cross-language consistency fixture for EvaluatorResponseVerifier (issue #44).
//
// Spec lists `evaluator_pass.jsonl` and `evaluator_fail.jsonl`; we follow the
// `fixtures/completion_check/sql_result.json` precedent of a single JSON file
// with a `cases` array.

type fixtureExpected struct {
	Kind     string `json:"kind"`
	Contains string `json:"contains"`
}

type fixtureCase struct {
	Name        string              `json:"name"`
	PassPattern string              `json:"pass_pattern"`
	FailPattern string              `json:"fail_pattern"`
	BuildResult sporecore.RunResult `json:"build_result"`
	EvalResult  sporecore.RunResult `json:"eval_result"`
	Expected    fixtureExpected     `json:"expected"`
}

type fixtureSuite struct {
	Cases []fixtureCase `json:"cases"`
}

// TestEvaluatorResponseVerifierFixtureReplay replays
// fixtures/verifier/evaluator_response.json case-for-case.
func TestEvaluatorResponseVerifierFixtureReplay(t *testing.T) {
	wd, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	// go/spore-core/verifier → ../../../fixtures/verifier/evaluator_response.json
	path := filepath.Join(wd, "..", "..", "..", "fixtures", "verifier", "evaluator_response.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var suite fixtureSuite
	if err := json.Unmarshal(raw, &suite); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	if len(suite.Cases) == 0 {
		t.Fatalf("fixture has no cases")
	}
	for _, c := range suite.Cases {
		c := c
		t.Run(c.Name, func(t *testing.T) {
			v, err := NewEvaluatorResponseVerifier(c.PassPattern, c.FailPattern, 3)
			if err != nil {
				t.Fatalf("case %q: regex compile: %v", c.Name, err)
			}
			in := &VerifierInput{
				BuildResult: c.BuildResult,
				EvalResult:  c.EvalResult,
				Workspace:   "/fixture",
				Iteration:   0,
			}
			got := v.Verify(context.Background(), in)
			switch c.Expected.Kind {
			case "passed":
				if got.Kind != VerdictPassed {
					t.Fatalf("case %q: expected Passed, got %+v", c.Name, got)
				}
			case "failed":
				if got.Kind != VerdictFailed {
					t.Fatalf("case %q: expected Failed, got %+v", c.Name, got)
				}
				if !strings.Contains(got.Reason, c.Expected.Contains) {
					t.Fatalf("case %q: expected reason to contain %q, got %q",
						c.Name, c.Expected.Contains, got.Reason)
				}
			default:
				t.Fatalf("case %q: unknown expected kind %q", c.Name, c.Expected.Kind)
			}
		})
	}
}

File: typescript/packages/core/tests/verifier-fixture-replay.test.ts
/**
 * Cross-language fixture replay for {@link EvaluatorResponseVerifier}
 * (spore-core issue #44). Loads the shared JSON fixture and asserts the
 * same outcomes the Rust suite asserts.
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { SessionId, verifier } from "../src/index.js";
import type { AggregateUsage, HaltReason, RunResult } from "../src/index.js";

const { EvaluatorResponseVerifier } = verifier;

interface FixtureUsage {
  input_tokens?: number;
  output_tokens?: number;
  cost_usd?: number;
}
interface FixtureSuccess {
  kind: "success";
  output: string;
  session_id: string;
  usage: FixtureUsage;
  turns: number;
}
interface FixtureFailure {
  kind: "failure";
  reason: HaltReason;
  session_id: string;
  usage: FixtureUsage;
  turns: number;
}
type FixtureRunResult = FixtureSuccess | FixtureFailure;

interface FixtureExpectedPassed {
  kind: "passed";
}
interface FixtureExpectedFailed {
  kind: "failed";
  contains: string;
}
type FixtureExpected = FixtureExpectedPassed | FixtureExpectedFailed;

interface FixtureCase {
  name: string;
  pass_pattern: string;
  fail_pattern: string;
  build_result: FixtureRunResult;
  eval_result: FixtureRunResult;
  expected: FixtureExpected;
}

interface FixtureFile {
  description?: string;
  cases: FixtureCase[];
}

const here = dirname(fileURLToPath(import.meta.url));
const fixturePath = resolve(here, "../../../../fixtures/verifier/evaluator_response.json");

function reifyUsage(u: FixtureUsage): AggregateUsage {
  return {
    input_tokens: u.input_tokens ?? 0,
    output_tokens: u.output_tokens ?? 0,
    cache_read_tokens: 0,
    cache_write_tokens: 0,
    cost_usd: u.cost_usd ?? 0,
  };
}

function reifyRunResult(r: FixtureRunResult): RunResult {
  if (r.kind === "success") {
    return {
      kind: "success",
      output: r.output,
      session_id: SessionId.of(r.session_id),
      usage: reifyUsage(r.usage),
      turns: r.turns,
    };
  }
  return {
    kind: "failure",
    reason: r.reason,
    session_id: SessionId.of(r.session_id),
    usage: reifyUsage(r.usage),
    turns: r.turns,
  };
}

describe("EvaluatorResponseVerifier fixture replay", () => {
  const text = readFileSync(fixturePath, "utf8");
  const file = JSON.parse(text) as FixtureFile;

  for (const c of file.cases) {
    it(c.name, async () => {
      const v = new EvaluatorResponseVerifier({
        pass_pattern: c.pass_pattern,
        fail_pattern: c.fail_pattern,
      });
      const verdict = await v.verify({
        build_result: reifyRunResult(c.build_result),
        eval_result: reifyRunResult(c.eval_result),
        workspace: "/fixture",
        iteration: 0,
      });
      if (c.expected.kind === "passed") {
        expect(verdict.kind).toBe("passed");
      } else {
        expect(verdict.kind).toBe("failed");
        if (verdict.kind === "failed") {
          expect(verdict.reason).toContain(c.expected.contains);
        }
      }
    });
  }
});
